src/app/gibbs/single_thread_sampler.cpp

```diff
@@ -157,25 +157,24 @@ inline int dd::SingleThreadSampler::draw_sample(Variable &variable,
       proposal = -1;
 
       // calculate potential for each proposal
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
+      for (size_t i = 0; i < variable.cardinality; i++) {
+        int propose = variable.get_domain(i);
         if (is_free_sample) {
-          varlen_potential_buffer[propose] =
+          varlen_potential_buffer[i] =
               p_fg->template potential<true>(variable, propose);
         } else {
-          varlen_potential_buffer[propose] =
+          varlen_potential_buffer[i] =
               p_fg->template potential<false>(variable, propose);
         }
-        sum = logadd(sum, varlen_potential_buffer[propose]);
+        sum = logadd(sum, varlen_potential_buffer[i]);
       }
 
       // flip a coin
       *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
-        acc += exp(varlen_potential_buffer[propose] - sum);
+      for (size_t i = 0; i < variable.cardinality; i++) {
+        acc += exp(varlen_potential_buffer[i] - sum);
         if (*this->p_rand_obj_buf <= acc) {
-          proposal = propose;
+          proposal = variable.get_domain(i);
```

Comment:
I think those lower-level details (dense/sparse) should be abstracted away from its callers. We should have a unified high-level interface. If we do it as above, we will have to repeat the same code whenever we iterate over the domain. If we support more types of variables in the future, these repeated code will all need to be modified...